### ISSUE TITLE
Strongly typed createParamDecorator arguments

### ISSUE THREAD

**Issue by Gregroam** [nestjs/nest#3356]
## Feature Request

When you create a parameter decorator, the arguments that it can accept aren't strongly typed. For example take this code snippet from the docs https://docs.nestjs.com/custom-decorators

```
import { createParamDecorator } from '@nestjs/common';

export const User = createParamDecorator((data: string, req) => {
  return data ? req.user && req.user[data] : req.user;
});
```

When using the `@User()` decorator I can pass any argument to it, it doesn't validate that its a  string.
Example `@User(100)` should be invalid but its not.

It would be nice to get a more strongly typed decorator for examples where you want to pass an options object and have intellisense on it.

**Comment by kamilmysliwiec** [comment]
Good idea. Would you like to create a PR?

**Comment by Gregroam** [comment]
Sure thing I can give it a go!

**Comment by galesky** [comment]
I've tackled this today, should work on it some more over the weekend. At this point i've made possible to define the typing through generics. 

You guys think this is a cool approach for a v1 ?

**Comment by Akxe** [comment]
This is quite crucial if it wants to be actually used...

**Comment by kamilmysliwiec** [comment]
Fixed in 6.11.0

**Comment by lamuertepeluda** [comment]
Guys this new feature caused me troubles in my decorator: I had my decorator accepting an object (a configuration, with a defined interface) and now it is not working anymore, i.e. the decorator receives only default params. I think this is a sort of regression.

In my case I was using Papa parse for a CSV upload decorator with a following AJV validation pipe

Something like

```ts
interface IPostDataUploadParams {
  dataFieldName: string // Name of the data (CSV file) field
  schemaFieldName?: string // Name of the schema (JSON encoded string) field,
  schemaURL?: string // URL of the schema (schemaFieldName takes precedence over this)
  truncateOptions?: {
    precision?: number
    coordinates?: number
    mutate?: boolean
  }
  dynamicTyping?:
    | boolean
    | { [headerName: string]: boolean; [columnNumber: number]: boolean }
    | ((field: string | number) => boolean)
  transform?(value: string, field: string | number): any
  transformHeader?(header: string): string
}

export const CSVUpload = createParamDecorator(
  async (
    params: IPostDataUploadParams = {
      dataFieldName: 'file',
      schemaFieldName: 'schema',
      dynamicTyping: true
    },
    req
  ): Promise<PostDataDto> => {
//... does stuff accessing params, like parsing a csv file, but now params is stuck to default values! (in prev version it was working)
}

// Controller
@Post('/upload-csv/')
@UseInterceptors(FileInterceptor('file'))
public async uploadCSVFile(
    @CSVUpload(
      {
        dataFieldName: 'file',
        schemaURL: SCHEMA_URL,
        dynamicTyping: false,
        transform: ()=>{ /* function that does stuff */}
      },
      new AjvValidationPipe<MyPipeArgumentType>()
    )
    values: PostDataDto
  ) {
//.. controller code
}


```
